### PR TITLE
Implement umfPool[Set/Get]Tag

### PR DESCRIPTION
**Pull request by igchor**
Implements https://github.com/oneapi-src/unified-memory-framework/issues/687

### Checklist

- [x] Code compiles without errors locally
- [x] All tests pass locally
- [x] CI workflows execute properly

- [ ] CI workflows, not executed per PR (e.g. Nightly), execute properly 
- [ ] New tests added, especially if they will fail without my changes
- [ ] Added/extended example(s) to cover this functionality
- [ ] Extended the README/documentation
- [ ] All newly added source files have a license
- [ ] All newly added source files are referenced in CMake files
- [ ] Logger (with debug/info/... messages) is used

### CODE REVIEW

**Review by ldorau — APPROVED** [review]

**Comment by bratpiorka** [comment]
@igchor do we need this for 0.10x / 2025.1?

File: include/umf/memory_pool.h [hunk 1]
```diff
@@ -170,6 +170,22 @@ umf_memory_pool_handle_t umfPoolByPtr(const void *ptr);
 umf_result_t umfPoolGetMemoryProvider(umf_memory_pool_handle_t hPool,
                                       umf_memory_provider_handle_t *hProvider);
 
+///
+/// @brief Set a custom tag on the memory pool that can be later retrieved using umfPoolGetTag.
+/// @param hPool specified memory pool
+/// @param tag tag to be set
+/// @param oldTag [out] previous tag set on the memory pool
```
Comment (lplewa) [inline include/umf/memory_pool.h:177]: document it that it is optional and can be bull

Comment (igchor) [reply include/umf/memory_pool.h]: Done.

**Comment by igchor** [comment]
> @igchor do we need this for 0.10x / 2025.1?

No, for now I only want to use it for v2 L0 adapter which will not be released as part of 2025.1

**Review by bratpiorka — APPROVED** [review]

File: test/memoryPoolAPI.cpp [hunk 2]
```diff
@@ -178,6 +178,107 @@ TEST_F(test, BasicPoolByPtrTest) {
...
+    ASSERT_TRUE(op == NUM_OPS_PER_THREAD - 1);
+}
+
+TEST_F(tagTest, SetAndGetInvalidPtr) {
+    umf_result_t ret = umfPoolSetTag(pool.get(), nullptr, nullptr);
+    ASSERT_EQ(ret, UMF_RESULT_SUCCESS);
+
+    ret = umfPoolGetTag(pool.get(), nullptr);
+    ASSERT_EQ(ret, UMF_RESULT_ERROR_INVALID_ARGUMENT);
+}
```
Comment (lukaszstolarczuk) [inline test/memoryPoolAPI.cpp:280]: I'm missing 2 tests for invalid hPool

File: test/memoryPoolAPI.cpp [hunk 3]
```diff
@@ -178,6 +178,107 @@ TEST_F(test, BasicPoolByPtrTest) {
...
+                    umfPoolSetTag(pool.get(), encodeTag(i, j), &oldTag);
+                ASSERT_EQ(ret, UMF_RESULT_SUCCESS);
+
+                void *queriedTag;
+                ret = umfPoolGetTag(pool.get(), &queriedTag);
+                ASSERT_EQ(ret, UMF_RESULT_SUCCESS);
+
+                auto [t1, op1] = decodeTag(oldTag);
+                auto [t2, op2] = decodeTag(queriedTag);
+                // if the tag was set by the same thread, the op part should same or higher
```
Comment (lukaszstolarczuk) [inline test/memoryPoolAPI.cpp:255]: `should be the same`

File: test/memoryPoolAPI.cpp [hunk 4]
```diff
@@ -178,6 +178,107 @@ TEST_F(test, BasicPoolByPtrTest) {
...
+    void *tag;
+    ret = umfPoolGetTag(pool.get(), &tag);
+    ASSERT_EQ(ret, UMF_RESULT_SUCCESS);
+    ASSERT_EQ(tag, (void *)0x99);
+
+    void *oldTag;
+    ret = umfPoolSetTag(pool.get(), (void *)0x100, &oldTag);
+    ASSERT_EQ(ret, UMF_RESULT_SUCCESS);
+    ASSERT_EQ(oldTag, (void *)0x99);
+}
```
Comment (lukaszstolarczuk) [inline test/memoryPoolAPI.cpp:206]: perhaps, add get once more...?

File: test/memoryPoolAPI.cpp (hunk 2 above)
Comment (igchor) [reply test/memoryPoolAPI.cpp:280]: Right, done.

File: test/memoryPoolAPI.cpp (hunk 3 above)
Comment (igchor) [reply test/memoryPoolAPI.cpp:255]: Done.

File: test/memoryPoolAPI.cpp (hunk 4 above)
Comment (igchor) [reply test/memoryPoolAPI.cpp:206]: Done

**Review by lukaszstolarczuk — APPROVED** [review]

**Pull request merged**